lib/command/src/Obelisk/Command/Upgrade.hs

```diff
@@ -0,0 +1,186 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module Obelisk.Command.Upgrade where
+
+import Control.Monad (forM_, unless, void)
+import Control.Monad.IO.Class (liftIO)
+import Control.Monad.Trans.Maybe (MaybeT (..), runMaybeT)
+import qualified Data.Map as Map
+import Data.Semigroup ((<>))
+import Data.Text (Text)
+import qualified Data.Text as T
+import System.Environment (getExecutablePath)
+import System.FilePath
+import System.Posix.Process (executeFile)
+import System.Process (proc)
+
+import Obelisk.App (MonadObelisk)
+import Obelisk.CliApp
+import Obelisk.Command.Utils
+
+import Obelisk.Command.Project (toImplDir)
+import Obelisk.Command.Project (findProjectObeliskCommand)
+import Obelisk.Command.Thunk (updateThunk)
+
+import Obelisk.Migration
+
+data MigrationGraph
+  = MigrationGraph_ObeliskUpgrade
+  | MigrationGraph_ObeliskHandoff
+  deriving (Eq, Show)
+
+graphName :: MigrationGraph -> Text
+graphName = \case
+  MigrationGraph_ObeliskHandoff -> "obelisk-handoff"
+  MigrationGraph_ObeliskUpgrade -> "obelisk-upgrade"
+
+fromGraphName :: Text -> MigrationGraph
+fromGraphName = \case
+  "obelisk-handoff" -> MigrationGraph_ObeliskHandoff
+  "obelisk-upgrade" -> MigrationGraph_ObeliskUpgrade
+  _ -> error "Invalid graph name specified"
+
+
+ensureCleanProject :: MonadObelisk m => FilePath -> m ()
+ensureCleanProject project = ensureCleanGitRepo project False "Cannot upgrade with uncommited changes"
+
+-- | Decide whether we (ambient ob) should handoff to project obelisk before performing upgrade
+decideHandOffToProjectOb :: MonadObelisk m => FilePath ->  m Bool
+decideHandOffToProjectOb project = do
+  ensureCleanProject project
+  updateThunk (toImplDir project) $ \projectOb -> do
+    getMigrationGraph projectOb MigrationGraph_ObeliskHandoff >>= \case
```

Comment:
I don't think we should be checking the project migration graph for anything.  What we discussed previously was using the ambient ob's migration graph; is there a reason that can't work?